File: data-structures/ring-buffer/implementations/ring-buffer.go
package implementations

import "dsa/data-structures/ring-buffer/interfaces"

type MyRingBuffer[T any] struct {
	initialCapacity int
	head            int
	tail            int
	currentLength   int
	container       []T
}

var _ interfaces.RingBuffer[int] = (*MyRingBuffer[int])(nil)

func NewMyRingBuffer[T any](initialCapacity int) *MyRingBuffer[T] {
	if initialCapacity <= 0 {
		panic("initialCapacity must be greater than zero")
	}

	return &MyRingBuffer[T]{
		initialCapacity: initialCapacity,
		container:       make([]T, initialCapacity),
		head:            0,
		tail:            0,
		currentLength:   0,
	}
}

func (b *MyRingBuffer[T]) Length() int {
	return b.currentLength
}

func (b *MyRingBuffer[T]) Capacity() int {
	return b.initialCapacity
}

func (b *MyRingBuffer[T]) IsEmpty() bool {
	return b.Length() <= 0
}

func (b *MyRingBuffer[T]) IsFull() bool {
	return b.Length() == b.Capacity()
}

func (b *MyRingBuffer[T]) Front() T {
	if b.IsEmpty() {
		panic("buffer is full")
	}

	return b.container[b.head]
}

func (b *MyRingBuffer[T]) Rear() T {
	if b.IsEmpty() {
		panic("buffer is full")
	}

	idx := (b.tail - 1 + b.Capacity()) % b.Capacity()
	return b.container[idx]
}

func (b *MyRingBuffer[T]) Enqueue(value T) bool {
	if b.IsFull() {
		return false
	}

	b.container[b.tail] = value
	b.currentLength++
	b.tail = (b.tail + 1) % b.Capacity()
	return true
}

func (b *MyRingBuffer[T]) Dequeue() T {
	if b.IsEmpty() {
		panic("dequeue: buffer is empty")
	}
	val := b.container[b.head]
	b.currentLength--
	b.head = (b.head + 1) % b.Capacity()

	return val
}

func (b *MyRingBuffer[T]) Clear() {
	b.currentLength = 0
	b.head = 0
	b.tail = 0
	b.container = nil
}

func (b *MyRingBuffer[T]) ToSlice() []T {
	rs := make([]T, 0, b.Length())

	for i := 0; i < b.Length(); i++ {
		idx := (b.head + i) % b.Capacity()
		rs = append(rs, b.container[idx])
	}

	return rs
}

func (b *MyRingBuffer[T]) InitialCapacity() int {
	return b.initialCapacity
}

File: data-structures/ring-buffer/implementations/ring-buffer.ts
import { RingBuffer } from "../interfaces/ring-buffer";

export class MyRingBuffer<T> implements RingBuffer<T> {
  private readonly initialCapacity: number;
  private container: Array<T>
  private head: number;
  private tail: number;

  // bookeeping
  private currentLength: number


  public constructor(initialCapacity = 4) {
    if (initialCapacity <= 0) {
      throw new Error("initialCapacity must be greater than zero");
    }

    this.initialCapacity = initialCapacity;
    this.container = new Array(initialCapacity);
    this.head = 0;
    this.tail = 0
    this.currentLength = 0

  }

  public length(): number {
    return this.currentLength
  }

  public capacity(): number {
    return this.initialCapacity
  }

  public isEmpty(): boolean {
    return this.length() <= 0
  }

  public isFull(): boolean {
    return this.length() === this.capacity()
  }

  public front(): T | undefined {
    if(this.isEmpty()) return undefined
    return this.container[this.head]
  }

  public rear(): T | undefined {
    if(this.isEmpty()) return undefined;
    let rearIndex = (this.tail - 1 + this.capacity()) % this.capacity()
    return this.container[rearIndex]
  }

  public enqueue(_value: T): boolean {
    if(this.isFull()) return false
    
    this.container[this.tail] = _value
    this.currentLength++
    this.tail = (this.tail + 1) % this.getInitialCapacity()

    return true
  }

  public dequeue(): T | undefined {
    if(this.isEmpty()) return undefined
    const val = this.container[this.head]
    this.head = (this.head + 1) % this.getInitialCapacity()
    this.currentLength--
    return val
  }

  public clear(): void {
    this.container = new Array(this.initialCapacity)
    this.currentLength = 0
  }

  public toArray(): T[] {
   let result = [] as T[]

   for(let i = 0; i < this.length(); i++) {
    let idx = (this.head + i) % this.capacity()
    const v = this.container[idx]
    if(v && v !== undefined) {
      result.push(v)
    }
   }

   return result
  }

  protected getInitialCapacity(): number {
    return this.initialCapacity;
  }
}

